### ISSUE TITLE
Consider building cp38 wheels [enhancement]

### ISSUE THREAD

**Issue by JoshuaRack** [piwheels/piwheels#242]
Python 3.8 has been around for a couple of years now, and judging by the number of blogs and forum posts on the topic I'm far from the only person using it on a pi. I realise it's a big ask, an up to date python 3.8 build would need to be maintained and it's obviously going to add significantly to the wheel building load, but even just building a subset (most popular 100?) of packages would be a big help to us 3.8 users and would help gauge how popular this might be.  

I'm far from a python or linux expert but I'd be more than happy to help if this seems viable.

**Comment by bennuttall** [comment]
> Python 3.8 has been around for a couple of years now

It was released 10 months ago: https://docs.python.org/3/whatsnew/3.8.html

Sorry, it's not as easy as that. Python 3.8 is not in Debian, so people using it are compiling it themselves. Depending on when they build it, they could have any minor version. And with 3.9 or other future versions, maybe even pre-release versions.

When a wheel is compiled for others to use, it's not as simple as matching the ABI. You have to provide the same shared libraries it was built with. When using a standard stable Debian build, that's manageable because everyone is using (and has access to) the same set of libraries we do. It's still [complex](https://blog.piwheels.org/how-to-work-out-the-missing-dependencies-for-a-python-package/), but it's manageable, and we have tools for providing people with the dependencies they need.

If we were to compile a version of Python 3.8 and build cp38m wheels, either we wouldn't have the right `python3-` apt packages providing libraries needed. It's complicated, and we wouldn't be able to solve it in a universal way that would work for everyone.

Next Summer, Debian Bullseye will be released with a newer version of Python and as soon as it's stable we'll start building for it.

In Q2 of this year, we had a very small proportion of users using Python 3.8. Unfortunately much smaller than 2.7 :(

![](https://blog.piwheels.org/wp-content/uploads/2020/07/py-vers.png)

**Comment by waveform80** [comment]
Yup, as Ben's noted: piwheels (or any copy of it) is quite tightly linked to the distro it's built against. At some point, when I've added the build-queue logic to handle multiple architectures we'll be moving onto adding aarch64 support for raspios, and I'll be looking at setting up an ubuntu variant for the supported releases. But this will all still be based on whatever version of python is shipped by default with the targetted distro.

In other words, roll-your-own python, and you get to roll-your-own wheels too.

**Comment by JoshuaRack** [comment]
Thanks for the quick and detailed responses . 

I don't know where I got the idea 3.8 had been around for a couple of years, perhaps I mixed up 3.7.8 and 3.8. I hadn't considered the shared library issue, it's clearly a deal breaker. 

I don't actually need 3.8, I just recently got the 2nd ed of Efficient Python which mentions a few 3.8 features so thought I'd give them a spin. After the long and slow road of building scipy etc I figured that it wasn't something I'd want to do again in a hurry so looked into building wheels to use for other projects, ended up here and thought if I was building them for myself then maybe they could work for everyone, hence the request.  

Many thanks for all the work on this, my foray into 3.8 has been a strong reminder of just how helpful this project is.
